src/thunderbots/software/ai/passing/pass_generator.cpp

```diff
@@ -0,0 +1,200 @@
+
+
+PassGenerator::PassGenerator(double min_reasonable_pass_quality):
+min_reasonable_pass_quality(min_reasonable_pass_quality),
+optimizer(optimizer_param_weights),
+best_known_pass(std::nullopt),
+in_destructor(false)
+{
+    // Generate the initial set of passes
+    passes_to_optimize = generatePasses(this->num_passes_to_optimize);
+
+    // Start the thread to do the pass generation in the background
+    // The lambda expression here is needed so that we can call
+    // `continuouslyGeneratePasses()`, which is not a static function
+    pass_generation_thread = std::thread([this](){ return continuouslyGeneratePasses(); });
+}
+
+void PassGenerator::setWorld(World world) {
+    // Take ownership of the world for the duration of this function
+    std::lock_guard<std::mutex> world_lock(world_mutex);
+
+    // Update the world
+    this->world = std::move(world);
+}
+
+void PassGenerator::setPasserPoint(Point passer_point) {
+    // Take ownership of the passer_point for the duration of this function
+    std::lock_guard<std::mutex> passer_point_lock(passer_point_mutex);
+
+    // Update the passer point
+    this->passer_point = passer_point;
+}
+
+std::optional<Pass> PassGenerator::getBestPassSoFar() {
+    // Take ownership of the best_known_pass for the duration of this function
+    std::lock_guard<std::mutex> best_known_pass_lock(best_known_pass_mutex);
+
+    return best_known_pass;
+}
+
+PassGenerator::~PassGenerator() {
+    // Set this flag so pass_generation_thread knows to end (also making sure to
+    // properly take and give ownership of the flag)
+    in_destructor_mutex.lock();
+    in_destructor = true;
+    in_destructor_mutex.unlock();
+
+    // Join to pass_generation_thread so that we wait for it to exit before destructing
+    // the thread object. If we do not wait for thread to finish executing, it will
+    // call `std::terminate` and kill our whole program when it finishes
+    pass_generation_thread.join();
+}
+
+void PassGenerator::continuouslyGeneratePasses() {
+    // Take ownership of the in_destructor flag so we can use it for the conditional
+    // check
+    in_destructor_mutex.lock();
+    while(!in_destructor){
+        // Give up ownership of the in_destructor flag now that we're done the
+        // conditional check
+        in_destructor_mutex.unlock();
+
+        optimizePasses();
+        pruneAndReplacePasses();
+        saveBestPath();
+
+        // Yield to allow other threads to run. This is particularly important if we
+        // have this thread and another running on one core
+        std::this_thread::yield();
+
+        // Take ownership of the `in_destructor` flag so we can use it for the conditional
+        // check
+        in_destructor_mutex.lock();
+    }
+}
+
+void PassGenerator::optimizePasses() {
+    // The objective function we minimize in gradient descent to improve each pass
+    // that we're optimizing
+    static const auto objective_function =
```

Comment:
Any reason this is `static`?

Reply:
because it doesn't changed? changed to just `const` to be safe